R/data.table.R

```diff
@@ -1337,21 +1337,21 @@ replace_dot_alias = function(e) {
     if (is.data.table(jval)) {
       setattr(jval, 'class', class(x)) # fix for #64
       # can jval be sorted by the same key as x? improved for #4498
-      get_shared_keys = function(jsub, jvnames, key) {
+      get_shared_keys = function(jsub, jvnames, sdvars, key) {
         if (is.null(key)) return(NULL)
-        if (!jsub %iscall% "list") return(NULL)
-        jnames = as.character(Filter(is.name, jsub)[-1L])
+        if (!((SD_only <- jsub == quote(.SD))|| jsub %iscall% "list")) return(NULL)
+        jnames = if (SD_only) sdvars else as.character(Filter(is.name, jsub)[-1L])
         key_idx = chmatch(key, jnames)
         missing_keys = which(is.na(key_idx))
         if (length(missing_keys) && missing_keys[1L] == 1L) return(NULL)
         if (!length(missing_keys)) return(jvnames[key_idx])
         return(jvnames[head(key_idx, missing_keys[1L] - 1L)])
       }
-      shared_keys = get_shared_keys(jsub, jvnames, key(x))
+      shared_keys = get_shared_keys(jsub, jvnames, sdvars = sdvars, key(x))
       if (is.null(irows) && !is.null(shared_keys)) {
         setattr(jval, 'sorted', shared_keys)
         # potentially inefficient backup -- check if jval is sorted by key(x)
-      } else if (haskey(x) && all(key(x) %chin% names(jval)) && suppressWarnings(is.sorted(jval, by=key(x)))) { # TO DO: perhaps this usage of is.sorted should be allowed internally then (tidy up and make efficient)
+      } else if (haskey(x) && all(key(x) %chin% names(jval)) && (is.null(irows) || (!roll && .Call(CisOrderedSubset, irows, nrow(x))) || suppressWarnings(is.sorted(jval, by=key(x))))) { # TO DO: perhaps this usage of is.sorted should be allowed internally then (tidy up and make efficient)
```

Comment:
hmm isn't this condition wrong though? Since all we've checked is that `key(x)` shows up in `names(jval)`, a query like this would pass right?

```
DT = data.table(a=1, b=2, c=3, key='a,b,c')
DT[ , .(a = 10:1, b=1:10, c=10:1)]
```

(1) `DT` is keyed
(2) `a,b,c` all show up as names in `jval`
(3) `is.null(irows)`

We still need to run the `is.sorted` in this case right?

Reply:
^ confirmed on this example

Reply:
@ColeMiller1 I guess we want this new condition inside the earlier branch right?

```
if ((is.null(irows) || (!roll && .Call(CisOrderedSubset, irows, nrow(x)))) && !is.null(shared_keys)) {
```

Reply:
It is a fragile approach. I wanted to capture the example of ```c(.SD,  .(newvar = 3+3))``` but based on Matt's work on is.sorted, I do not think we need NSE to try to capture it. Will remove now.